src/server/tran_server.cpp

```diff
@@ -125,17 +125,44 @@ tran_server::boot (const char *db_name)
       return error_code;
     }
 
+  /*
+    * At least one PS is given by the configuration.
+    * Even if uses_remote_storage () == false, the remote storage can exist.
+    */
   if (m_page_server_conn_vec.empty () == false)
     {
-      /*
-       * At least one PS is given by the configuration.
-       * Even if uses_remote_storage () == false, the remote storage can exist.
-       */
-      error_code = reset_main_connection ();
-      if (error_code != NO_ERROR)
+      const auto start_time = std::chrono::steady_clock::now ();
+      while (true)
 	{
-	  assert (false);
-	  return error_code;
+	  error_code = reset_main_connection ();
+	  if (error_code == NO_ERROR)
+	    {
+	      break;
+	    }
+	  else
+	    {
+	      /* TEMPORARY waiting and timeout
+	      *
+	      * TODO: Remove this and just make sure reset_main_connection doesn't fail
+	      *       when the ATS recovery with handshakes with multiple page servers comes in.
+	      *
+	      * For now, the main connection may not be able to be set for a while. It can be after one it set to CONNECTED state.
+	      * When the handshakes comes in, it's guaranteed that at least one connection is completely CONNECTED here
+	      * and it will be the main connection. Until then, we just waits here until a conenction is ready.
+	      */
+	      const auto current_time = std::chrono::steady_clock::now ();
+	      const auto duration = std::chrono::duration_cast<std::chrono::seconds> (current_time - start_time).count();
+	      if (duration > 30) // timeout: 30 seconds
```

Comment:
Does it make sense to wait at this point? Since m_ps_connector.start(); is called outside the while loop, waiting doesn't seem to work that you exepected.

Reply:
It doesn't wait for the `m_ps_connector` to connect. All connections have already been established in `init_page_server_hosts` above. However, the state of them could be "CONNECTING", which can't be selected as the main connection. It will be set to CONNECTED when a PS reply that its catchup is done.

Reply:
I thought a case that all connect() call returns the ER_NET_PAGESERVER_CONNECTION error.

Reply:
You mean all `connect()` in `init_page_server_hosts()` above, right? In this case, the error is returned below calling the function. This loop is not reached.

Reply:
Oh. I got it.